pilot/pkg/model/push_context.go

```diff
@@ -1588,26 +1587,33 @@ func (ps *PushContext) initServiceAccounts(env *Environment, services []*Service
 			if port.Protocol == protocol.UDP {
 				continue
 			}
+
 			var accounts sets.String
 			func() {
 				// First get endpoint level service accounts
 				shard, f := env.EndpointIndex.ShardsForService(string(svc.Hostname), svc.Attributes.Namespace)
 				if f {
 					shard.RLock()
-					defer shard.RUnlock()
-					accounts = shard.ServiceAccounts
+					// copy here to reduce the lock time, the copy cost can be ignored
+					// endpoints could update frequently, so the longer it locks, the more likely it will block other threads.
+					accounts = shard.ServiceAccounts.Copy()
+					shard.RUnlock()
 				}
 				if len(svc.ServiceAccounts) > 0 {
-					accounts = accounts.Copy().InsertAll(svc.ServiceAccounts...)
+					if accounts == nil {
+						accounts = sets.New(svc.ServiceAccounts...)
+					} else {
+						accounts = accounts.InsertAll(svc.ServiceAccounts...)
+					}
 				}
 				sa := sets.SortedList(spiffe.ExpandWithTrustDomains(accounts, ps.Mesh.TrustDomainAliases))
 				key := serviceAccountKey{
 					hostname:  svc.Hostname,
 					namespace: svc.Attributes.Namespace,
-					port:      port.Port,
```

Comment:
Canwe remove the entire loop over ports now?

Reply:
sure, can do it